From d4d0c1b1bea98c251a57a13784641454bb811c1a Mon Sep 17 00:00:00 2001
From: mangadi <mangadi@qti.qualcomm.com>
Date: Wed, 18 Dec 2024 18:28:29 +0530
Subject: [PATCH] Add VkSharedPresentSurfaceCapabilitiesKHR to
 vkGetPhysicalDeviceSurfaceCapabilities2KHR

which is required by VK_KHR_shared_presentable_image
also expose VK_KHR_get_surface_capabilities2 as supported
extention from loader

Signed-off-by: Mahesh Angadi <mangadi@qti.qualcomm.com>
Upstream-Status: Pending

diff --git a/loader/loader.c b/loader/loader.c
index 7470790aa..bc77312fe 100644
--- a/loader/loader.c
+++ b/loader/loader.c
@@ -1224,6 +1224,14 @@ VkResult loader_get_icd_loader_instance_extensions(const struct loader_instance
         goto out;
     }

+    // Add VK_KHR_GET_SURFACE_CAPABILITIES_2
+    static const VkExtensionProperties loader_internal_extension_info[] = {
+        {VK_KHR_GET_SURFACE_CAPABILITIES_2_EXTENSION_NAME, VK_KHR_GET_SURFACE_CAPABILITIES_2_SPEC_VERSION},
+    };
+
+    loader_add_to_ext_list(inst, inst_exts, sizeof(loader_internal_extension_info) / sizeof(VkExtensionProperties),
+                           loader_internal_extension_info);
+
 out:
     return res;
 }
diff --git a/loader/wsi.c b/loader/wsi.c
index 46c81b9ee..d9aa9a5e6 100644
--- a/loader/wsi.c
+++ b/loader/wsi.c
@@ -2675,11 +2675,20 @@ VKAPI_ATTR VkResult VKAPI_CALL terminator_GetPhysicalDeviceSurfaceCapabilities2K
         VkResult res = icd_term->dispatch.GetPhysicalDeviceSurfaceCapabilitiesKHR(phys_dev_term->phys_dev, surface,
                                                                                   &pSurfaceCapabilities->surfaceCapabilities);

-        if (pSurfaceCapabilities->pNext != NULL) {
-            loader_log(icd_term->this_instance, VULKAN_LOADER_WARN_BIT, 0,
-                       "vkGetPhysicalDeviceSurfaceCapabilities2KHR: Emulation found unrecognized structure type in "
-                       "pSurfaceCapabilities->pNext - this struct will be ignored");
+        VkBaseOutStructure *pNext = (VkBaseOutStructure *)pSurfaceCapabilities->pNext;
+        while (pNext != NULL) {
+            if ((int)pNext->sType == VK_STRUCTURE_TYPE_SHARED_PRESENT_SURFACE_CAPABILITIES_KHR) {
+                // Update usage flags for shared presentable image surface
+                VkImageUsageFlags surfaceUsageFlags = pSurfaceCapabilities->surfaceCapabilities.supportedUsageFlags;
+                ((VkSharedPresentSurfaceCapabilitiesKHR*)(pSurfaceCapabilities->pNext))->sharedPresentSupportedUsageFlags = surfaceUsageFlags;
+            } else {
+                loader_log(icd_term->this_instance, VK_DEBUG_REPORT_WARNING_BIT_EXT, 0,
+                       "vkGetPhysicalDeviceSurfaceCapabilities2KHR: Emulation found unrecognized structure type 0x%x in "
+                       "pSurfaceCapabilities->pNext - this struct will be ignored", pNext->sType);
+            }
+            pNext = (VkBaseOutStructure *)pNext->pNext;
         }
+
         return res;
     }
 }
--
2.25.1

